File: py/problem.py
class Problem:
    '''Class that holds and calculates problem specific information'''
    def __init__(self, state_space, start, goal):
        self.states = state_space
        self.initial = start
        self.goal = goal

    def actions(self, state):
        '''Find current state in state space then return that states actions'''
        for sp_state in self.states:
            if (sp_state['state']['x'] == state[0] and
                sp_state['state']['y'] == state[1]):
                return sp_state['actions']
            
    def action_cost(self, state1, state2):
        '''Calculate and return distance between current node and next node'''
        return ((abs(state1[0] - state2[0])**2) +
                (abs(state1[1] - state2[1])**2))**0.5

    def is_goal(self, state):
        '''Return whether the current state and goal state match'''
        if (state[0] == self.goal[0] and state[1] == self.goal[1]):
            return True
        else: return False

    def result(self, action):
        '''Return next state(dest) after taking action in current state'''
        return (action['dest']['x'], action['dest']['y'])

File: js/problem.mjs
class Problem {
    // Class that holds and calculates problem specific information
    constructor(stateSpace, start, goal) {
        this.states = stateSpace;
        this.initial = JSON.stringify(start);
        this.goal = goal;
    }
    actions(strState) {
        // Find current state in state space then return that states actions
        const state = JSON.parse(strState);
        for (let spState of this.states) {
            if (spState['state']['x'] === state['x'] &&
                spState['state']['y'] === state['y']) {
                return spState['actions'];
            }
        }
    }
    actionCost(strState1, strState2) {
        // Calculate and return distance between current node and next node
        const state1 = JSON.parse(strState1), state2 = JSON.parse(strState2);
        return Math.hypot(Math.abs(state1['x'] - state2['x']) +
                          Math.abs(state1['y'] - state2['y']));
    }
    isGoal(strState) {
        // Return whether the current state and goal state match
        const state = JSON.parse(strState);
        if (state['x'] === this.goal['x'] && state['y'] === this.goal['y']) {
            return true;
        }
        else {return false;}
    }
    result(action) {
        // Return next state(dest) after taking action in current state
        return JSON.stringify({'x': action['x'],
                               'y': action['y']});
    }
}

export { Problem };